libyul/optimiser/FunctionSpecializer.cpp

```diff
@@ -0,0 +1,157 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/FunctionSpecializer.h>
+
+#include <libyul/optimiser/ASTCopier.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDispenser.h>
+
+#include <libyul/AST.h>
+#include <libyul/YulString.h>
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/algorithm/any_of.hpp>
+
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+FunctionSpecializer::LiteralArguments FunctionSpecializer::specializableArguments(
+	FunctionCall const& _f
+)
+{
+	auto heuristic = [&](Expression const& _e) -> optional<Expression>
+	{
+		if (holds_alternative<Literal>(_e))
+			return ASTCopier{}.translate(_e);
+		return nullopt;
+	};
+
+	return applyMap(_f.arguments, heuristic);
+}
+
+void FunctionSpecializer::operator()(FunctionCall& _f)
+{
+	ASTModifier::operator()(_f);
+
+	// When backtracking is implemented, the restriction of recursive functions can be lifted.
+	if (
+		m_dialect.builtin(_f.functionName.name) ||
+		m_recursiveFunctions.count(_f.functionName.name)
+	)
+		return;
+
+	LiteralArguments arguments = specializableArguments(_f);
+
+	if (ranges::any_of(arguments, [](auto& _a) { return _a.has_value(); }))
+	{
+		YulString oldName = move(_f.functionName.name);
+		auto newName = m_nameDispenser.newName(oldName);
+
+		m_oldToNewMap[oldName].emplace_back(make_pair(newName, arguments));
+
+		_f.functionName.name = newName;
+		_f.arguments = util::filter(
+			_f.arguments,
+			applyMap(arguments, [](auto& _a) { return !_a; })
+		);
+	}
+}
+
+FunctionDefinition FunctionSpecializer::specialize(
+	FunctionDefinition const& _f,
+	YulString _newName,
+	FunctionSpecializer::LiteralArguments _arguments
+)
+{
+	yulAssert(_arguments.size() == _f.parameters.size(), "");
+
+	map<YulString, YulString> translatedNames = applyMap(
+		NameCollector{_f, NameCollector::OnlyVariables}.names(),
+		[&](auto& _name) -> pair<YulString, YulString>
+		{
```

Comment:
I usually put `{` in the previous line in lambdas like this, but fine with either.